### DISCUSSION TITLE
How to refresh incrementally

### DISCUSSION THREAD

**Discussion by merlinarer** [zauberzeug/nicegui#2772] [Q&A] [answered]
### Question

Hello there, in the example [chat_with_ai](https://github.com/zauberzeug/nicegui/tree/main/examples/chat_with_ai) , the refresh decorator destroys and recreate all elements every time it is called. This is really inefficient, especially when there is a lot of chat messeges in streaming style. I'm trying to modify the refresh function so that it only recreates the latest messege each time, achieving an incremental refresh effect. However, `self.targets` in `ui.refreshable` seems not to correspond to the actual elements.
```py
class CustomRefreshable(ui.refreshable):
    def prune(self) -> None:
        self.targets = [
            target
            for target in self.targets
            if target.container.client.id in Client.instances and target.container.id in target.container.client.elements
        ][-1:]
```

**Comment by falkoschindler** [comment]
Hi @merlinarer,

You're right, the use of `@refreshable` in the "Chat with AI" example is not ideal. For a real-world app you would probably need to add and update the individual chat and spinner elements yourself. I'm not sure if an "incremental refreshable" is the right approach, since it kind of contradicts its original idea of always clearing and re-creating everything. Instead I would simply add messages when they are generated and toggle the visibility of a spinner element depending on the `thinking` state, maybe via binding.

**Comment by merlinarer** [comment, marked as answer]
Thanks @falkoschindler , I finally got it worked. Here is an efficient class for "Chat with AI", and it is very easy to improve the original project just with few code to change.
```py
class CustomChatMessage:
    '''
    # user_avatar
    # bot_avatar
    # cls: main class
    '''
    def __init__(self, user_avatar, bot_avatar, cls):
        self.user_avatar = user_avatar
        self.bot_avatar = bot_avatar
        self.pointer = 0
        self.last_message = None
        self.spinner = None
        self.cls = cls

    def chat(self, name, text):
        with self.container:
            self.last_message = ui.chat_message(text=text, name=name, sent=name == 'You',
                                              avatar=self.user_avatar if name == 'You' else self.bot_avatar,
                                              text_html=True)
            self.spinner = ui.spinner(size='3rem').classes('self-center')
            self.spinner.bind_visibility_from(self.cls, "thinking")
            ui.run_javascript('document.getElementById("c4").scrollTo(0, document.getElementById("c4").scrollHeight)')

    def add_container(self, container):
        self.container = container

    def refresh(self, messeges):
        if len(messeges):
            name, text = messeges[-1]
            if self.spinner:
                self.spinner.delete()
            if len(messeges) > self.pointer:
                self.chat(name, text)
                self.pointer += 1
            else:
                if self.last_message:
                    self.last_message.delete()
                self.chat(name, text)
```

**Comment by rodja** [comment]
Interesting. Maybe this is worth to be integrated into the example via pull request?

**Reply by merlinarer** [reply]
hi, I create a pull request [here](https://github.com/zauberzeug/nicegui/pull/2800)
